File: performance-investigation/comparison/Benchmark.java
import java.util.Arrays;

public final class Benchmark {
    private static long fibonacci(long n) {
        if (n < 2) {
            return n;
        }
        return fibonacci(n - 1) + fibonacci(n - 2);
    }

    private static long sieve(int limit) {
        boolean[] isPrime = new boolean[limit + 1];
        Arrays.fill(isPrime, true);
        isPrime[0] = false;
        isPrime[1] = false;
        for (int prime = 2; (long) prime * prime <= limit; prime++) {
            if (isPrime[prime]) {
                for (int multiple = prime * prime; multiple <= limit; multiple += prime) {
                    isPrime[multiple] = false;
                }
            }
        }
        long count = 0;
        for (int value = 2; value <= limit; value++) {
            if (isPrime[value]) {
                count++;
            }
        }
        return count;
    }

    private static long matrixMultiply(int size) {
        int total = size * size;
        long[] a = new long[total];
        long[] b = new long[total];
        long[] c = new long[total];
        for (int index = 0; index < total; index++) {
            int row = index / size;
            int col = index - row * size;
            a[index] = row + col + 1L;
            b[index] = row * 2L + col + 1L;
        }
        for (int row = 0; row < size; row++) {
            for (int col = 0; col < size; col++) {
                long sum = 0;
                for (int k = 0; k < size; k++) {
                    sum += a[row * size + k] * b[k * size + col];
                }
                c[row * size + col] = sum;
            }
        }
        return c[0] + c[total - 1];
    }

    private static void swap(long[] values, int left, int right) {
        long value = values[left];
        values[left] = values[right];
        values[right] = value;
    }

    private static int partition(long[] values, int low, int high) {
        int middle = low + (high - low) / 2;
        swap(values, middle, high);
        long pivot = values[high];
        int store = low;
        for (int index = low; index < high; index++) {
            if (values[index] < pivot) {
                swap(values, index, store);
                store++;
            }
        }
        swap(values, store, high);
        return store;
    }

    private static void quicksort(long[] values, int low, int high) {
        if (low < high) {
            int split = partition(values, low, high);
            quicksort(values, low, split - 1);
            quicksort(values, split + 1, high);
        }
    }

    private static long quicksortBenchmark(int length) {
        long[] values = new long[length];
        for (int index = 0; index < length; index++) {
            values[index] = length - index;
        }
        quicksort(values, 0, length - 1);
        return values[0] + values[length / 2] + values[length - 1];
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            throw new IllegalArgumentException("usage: Benchmark <fib|sieve|matrix|quicksort> <size>");
        }
        String algorithm = args[0];
        int size = Integer.parseInt(args[1]);
        long result = switch (algorithm) {
            case "fib" -> fibonacci(size);
            case "sieve" -> sieve(size);
            case "matrix" -> matrixMultiply(size);
            case "quicksort" -> quicksortBenchmark(size);
            default -> throw new IllegalArgumentException("unknown algorithm: " + algorithm);
        };
        System.out.println(result);
    }
}

File: performance-investigation/comparison/benchmark.ml
let rec fibonacci n =
  if n < 2 then n else fibonacci (n - 1) + fibonacci (n - 2)

let sieve limit =
  let is_prime = Bytes.make (limit + 1) '\001' in
  Bytes.set is_prime 0 '\000';
  Bytes.set is_prime 1 '\000';
  let prime = ref 2 in
  while !prime * !prime <= limit do
    if Bytes.get is_prime !prime <> '\000' then begin
      let multiple = ref (!prime * !prime) in
      while !multiple <= limit do
        Bytes.set is_prime !multiple '\000';
        multiple := !multiple + !prime
      done
    end;
    incr prime
  done;
  let count = ref 0 in
  for value = 2 to limit do
    if Bytes.get is_prime value <> '\000' then incr count
  done;
  !count

let matrix_multiply size =
  let total = size * size in
  let a = Array.make total 0 in
  let b = Array.make total 0 in
  let c = Array.make total 0 in
  for index = 0 to total - 1 do
    let row = index / size in
    let col = index - (row * size) in
    a.(index) <- row + col + 1;
    b.(index) <- (row * 2) + col + 1
  done;
  for row = 0 to size - 1 do
    for col = 0 to size - 1 do
      let sum = ref 0 in
      for k = 0 to size - 1 do
        sum := !sum + (a.((row * size) + k) * b.((k * size) + col))
      done;
      c.((row * size) + col) <- !sum
    done
  done;
  c.(0) + c.(total - 1)

let swap values left right =
  let value = values.(left) in
  values.(left) <- values.(right);
  values.(right) <- value

let partition values low high =
  let middle = low + ((high - low) / 2) in
  swap values middle high;
  let pivot = values.(high) in
  let store = ref low in
  for index = low to high - 1 do
    if values.(index) < pivot then begin
      swap values index !store;
      incr store
    end
  done;
  swap values !store high;
  !store

let rec quicksort values low high =
  if low < high then begin
    let split = partition values low high in
    quicksort values low (split - 1);
    quicksort values (split + 1) high
  end

let quicksort_benchmark length =
  let values = Array.init length (fun index -> length - index) in
  quicksort values 0 (length - 1);
  values.(0) + values.(length / 2) + values.(length - 1)

let () =
  if Array.length Sys.argv <> 3 then
    failwith "usage: benchmark <fib|sieve|matrix|quicksort> <size>";
  let algorithm = Sys.argv.(1) in
  let size = int_of_string Sys.argv.(2) in
  let result =
    match algorithm with
    | "fib" -> fibonacci size
    | "sieve" -> sieve size
    | "matrix" -> matrix_multiply size
    | "quicksort" -> quicksort_benchmark size
    | _ -> failwith ("unknown algorithm: " ^ algorithm)
  in
  Printf.printf "%d\n" result
